src/rpc_ctx.c

```diff
@@ -97,70 +98,74 @@ rpc_ctx_alloc(CLIENT *clnt, struct timeval timeout)
-			__func__, ctx->xid, clnt);
+			"%s: %p fd %d call ctx insert failed xid %" PRIu32,
+			__func__, &rec->xprt, rec->xprt.xp_fd, ctx->xid);
 		rpc_ctx_release(ctx);
 		rpc_dplx_rui(rec);
 		return (NULL);
 	}
 	return (ctx);
 }
 
-/* RPC_CTX_FLAG_LOCKED, RPC_DPLX_FLAG_LOCKED
- */
-bool
-rpc_ctx_next_xid(rpc_ctx_t *ctx)
-{
-	CLIENT *clnt = ctx->ctx_u.clnt.clnt;
-	struct cx_data *cx = CX_DATA(clnt);
-	struct rpc_dplx_rec *rec = cx->cx_rec;
-
-	/* the lock protects both xid and rbtree */
-	opr_rbtree_remove(&rec->call_replies, &ctx->node_k);
-	ctx->xid = ++(rec->call_xid);
-	ctx->flags = RPC_CTX_FLAG_NONE;
-
-	if (opr_rbtree_insert(&rec->call_replies, &ctx->node_k)) {
-		__warnx(TIRPC_DEBUG_FLAG_ERROR,
-			"%s: call ctx insert failed (xid %d client %p)",
-			__func__, ctx->xid, clnt);
-		rpc_ctx_release(ctx);
-		rpc_dplx_rui(rec);
-		return (false);
-	}
-	return (true);
-}
-
 /* RPC_DPLX_FLAG_LOCKED
  */
-bool
-rpc_ctx_xfer_replymsg(struct svc_vc_xprt *xd, struct rpc_msg *msg)
+enum xprt_stat
+rpc_ctx_xfer_replymsg(struct svc_req *req)
 {
+	XDR *xdrs = req->rq_xdrs;
+	SVCXPRT *xprt = req->rq_xprt;
+	struct rpc_dplx_rec *rec = REC_XPRT(xprt);
 	rpc_ctx_t ctx_k, *ctx;
 	struct opr_rbtree_node *nv;
-	struct rpc_dplx_rec *rec = &xd->sx_dr;
-	rpc_dplx_lock_t *lk = &rec->recv.lock;
 
-	ctx_k.xid = msg->rm_xid;
+	ctx_k.xid = req->rq_msg.rm_xid;
 	nv = opr_rbtree_lookup(&rec->call_replies, &ctx_k.node_k);
-	if (nv) {
-		ctx = opr_containerof(nv, rpc_ctx_t, node_k);
-		atomic_set_uint16_t_bits(&ctx->flags, RPC_CTX_FLAG_ACKSYNC);
-		atomic_inc_uint32_t(&ctx->refcount);
-		ctx->cc_msg = *msg;	/* and stash reply header */
+	if (!nv) {
+		/* release internal locks */
+		return SVC_STAT(xprt);
+	}
 
-		/* signal the specific ctx  */
-		cond_signal(&ctx->we.cv);
+	ctx = opr_containerof(nv, rpc_ctx_t, node_k);
+
+	_seterr_reply(&req->rq_msg, &(ctx->error));
+	if (ctx->error.re_status == RPC_SUCCESS) {
+		if (!AUTH_VALIDATE(ctx->cc_auth,
+				   &(ctx->cc_msg.RPCM_ack.ar_verf))) {
+			ctx->error.re_status = RPC_AUTHERROR;
+			ctx->error.re_why = AUTH_INVALIDRESP;
+		} else if (ctx->cc_xdr.proc
+			   && !AUTH_UNWRAP(ctx->cc_auth, xdrs,
+					   ctx->cc_xdr.proc,
+					   ctx->cc_xdr.where)) {
+			if (ctx->error.re_status == RPC_SUCCESS)
+				ctx->error.re_status = RPC_CANTDECODERES;
+		}
+		ctx->refreshes = 0;
+	} else if (ctx->refreshes-- > 0
+		   && AUTH_REFRESH(ctx->cc_auth, &(ctx->cc_msg))) {
+		/* maybe our credentials need to be refreshed ... */
+		opr_rbtree_remove(&rec->call_replies, &ctx->node_k);
+		ctx->xid = ++(rec->call_xid);
+
+		if (opr_rbtree_insert(&rec->call_replies, &ctx->node_k)) {
+			__warnx(TIRPC_DEBUG_FLAG_ERROR,
+				"%s: %p fd %d call ctx insert failed xid %" PRIu32,
+				__func__, xprt, xprt->xp_fd, ctx->xid);
+			ctx->error.re_status = RPC_TLIERROR;
+			ctx->refreshes = 0;
+		}
+	}
+	atomic_set_uint16_t_bits(&ctx->flags, RPC_CTX_FLAG_ACKSYNC);
 
-		/* now, we must ourselves wait for the other side to run */
```

Comment:
this is conceptually great, but for verification we need to use the callback simulator or otherwise verify that callbacks work